### PR TITLE
[flutter_tools] add --null-assertions flag for debugging with null safety

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

Make the flag available as an opt-in. A follow up PR will enable this for the flutter CI and fix the known test failures. This will insert additional null checks between null-safe and on-null safe libraries, with then intention of catching violations before all code has migrated to null safety

This flag must be off by default because it breaks a non-trivial amount of existing code

https://github.com/flutter/flutter/issues/61042

### CODE REVIEW

File: packages/flutter_tools/lib/src/device.dart [hunk 1]
```diff
@@ -993,3 +997,12 @@ class NoOpDevicePortForwarder implements DevicePortForwarder {
   @override
   Future<void> dispose() async { }
 }
+
+String computeDartVmFlags(DebuggingOptions debuggingOptions) {
+  String flags = debuggingOptions.dartFlags ?? '';
+  if (debuggingOptions.nullAssertions) {
+    final String separator = flags.isEmpty ? '' : ',';
+    flags += '$separator--null_assertions';
```
Comment (jmagman) [inline packages/flutter_tools/lib/src/device.dart:1005]: Is making a list and `join`ing too clunky?

File: packages/flutter_tools/test/general.shard/ios/ios_device_start_prebuilt_test.dart [hunk 2]
```diff
@@ -319,7 +319,7 @@ void main() {
             '--disable-service-auth-codes',
             '--observatory-port=60700',
             '--start-paused',
-            '--dart-flags="--foo"',
+            '--dart-flags="--foo,--null_assertions"',
```
Comment (jmagman) [inline packages/flutter_tools/test/general.shard/ios/ios_device_start_prebuilt_test.dart:322]: Is there a test that's just `--null_assertions` without a `--foo`?

File: packages/flutter_tools/lib/src/device.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/device.dart:1005]: Very difficult for my sleepy brain 🤣

Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/device.dart:1005]: I just wanted you to know that I understand the fence post problem

File: packages/flutter_tools/test/general.shard/ios/ios_device_start_prebuilt_test.dart (hunk 2 above)
Comment (jonahwilliams) [reply packages/flutter_tools/test/general.shard/ios/ios_device_start_prebuilt_test.dart:322]: I added a unit test that tests various flag combinations in computeDartVmFlags

File: packages/flutter_tools/lib/src/device.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/device.dart]: fixed

File: packages/flutter_tools/lib/src/runner/flutter_command.dart [hunk 3]
```diff
@@ -500,6 +501,12 @@ abstract class FlutterCommand extends Command<void> {
       defaultsTo: null,
       hide: hide,
     );
+    argParser.addFlag(FlutterOptions.kNullAssertions,
+      help:
+        'Whether to perform additional null checks on the boundaries of null safe and '
```
Comment (jmagman) [inline packages/flutter_tools/lib/src/runner/flutter_command.dart:506]: Maybe `'Perform additional null checks...`

**Review by jmagman — APPROVED** [review]
LGTM

**Pull request merged**